eZ/Publish/Core/Repository/RoleService.php

```diff
@@ -183,6 +185,73 @@ public function createRoleDraft(APIRole $role): APIRoleDraft
         return $this->roleDomainMapper->buildDomainRoleDraftObject($spiRole);
     }
 
+    /**
+     * Copies an existing Role.
+     *
+     * @since 8.0
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to copy a Role
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException if the name of the role already exists or if limitation of the same type
+     *         is repeated in the policy create struct or if limitation is not allowed on module/function
+     * @throws \eZ\Publish\API\Repository\Exceptions\LimitationValidationException if a policy limitation in the $roleCopyStruct is not valid
+     *
+     * @param \eZ\Publish\API\Repository\Values\User\Role $role
+     * @param \eZ\Publish\API\Repository\Values\User\RoleCopyStruct $roleCopyStruct
+     *
+     * @return \eZ\Publish\API\Repository\Values\User\Role
+     */
+    public function copyRole(APIRole $role, APIRoleCopyStruct $roleCopyStruct): APIRole
+    {
+        if (!is_string($roleCopyStruct->newIdentifier) || empty($roleCopyStruct->newIdentifier)) {
+            throw new InvalidArgumentValue('newIdentifier', $roleCopyStruct->newIdentifier, 'RoleCopyStruct');
+        }
+
+        if (!$this->permissionResolver->canUser('role', 'create', $roleCopyStruct)) {
+            throw new UnauthorizedException('role', 'create');
+        }
+
+        try {
+            $existingRole = $this->loadRoleByIdentifier($roleCopyStruct->newIdentifier);
+
+            throw new InvalidArgumentException(
+                '$roleCopyStruct',
+                "Role '{$existingRole->id}' with the specified identifier '{$roleCopyStruct->newIdentifier}' " .
+                'already exists'
+            );
+        } catch (APINotFoundException $e) {
+            // Do nothing
+        }
+
+        foreach ($role->getPolicies() as $policy) {
+            $policyCreateStruct = new PolicyCreateStruct([
+                'module' => $policy->module,
+                'function' => $policy->function,
+            ]);
+            foreach ($policy->getLimitations() as $limitation) {
+                $policyCreateStruct->addLimitation($limitation);
+            }
+            $roleCopyStruct->addPolicy($policyCreateStruct);
+        }
+
+        $limitationValidationErrors = $this->validateRoleCopyStruct($roleCopyStruct);
+        if (!empty($limitationValidationErrors)) {
+            throw new LimitationValidationException($limitationValidationErrors);
+        }
+
+        $spiRoleCopyStruct = $this->roleDomainMapper->buildPersistenceRoleCopyStruct($roleCopyStruct, $role->id);
+
+        $this->repository->beginTransaction();
+        try {
+            $spiRole = $this->userHandler->copyRole($spiRoleCopyStruct);
+            $this->repository->commit();
+        } catch (Exception $e) {
+            $this->repository->rollback();
+            throw $e;
```

Comment:
API should throw API exceptions only. So in the worst case, if an unknown exception is thrown, it should be wrapped with `BadStateException`. Please make sure that PhpStorm with its static analysis doesn't complain about missing throws in PhpDoc.